### PR TITLE
allow the table name to be returned in SQL mapper

### LINKED ISSUE #767: Unable to recover table name of mapped sql table
It is currently not possible to return the table name of a mapped SQL table which makes comparing two mapper objects computationally needlessly expensive and potentially flawed (you could compare the fields for example, but can't guarantee they are for the same table in actuality in edge cases where you have two tables with the same field names)

### PR DESCRIPTION
**Pull request by JeromeGill**
fixes #767

### CODE REVIEW

**Comment by ikkez** [comment]
you are talking about the field names, but your code returns the table name. this is confusing.

**Comment by JeromeGill** [comment]
Away with the fairies today, corrected the term

**Comment by JeromeGill** [comment]
Goes with Goes with bcosca/fatfree-core#19

**Comment by JeromeGill** [comment]
Example use case:

```php
class UserEntity extends BaseEntity {

    public function __construct(Mapper $table = null){
        if('users' === $table->table()){
            parent::__construct($table);
        } else {
            parent::__construct('users');
        }
    }

class BaseEntity {
    public function __construct($table){
        if($table instanceof Mapper){
            $this->model = $table;
        } else {
            $f3 = \Base::instance();
            $db = $f3->get('DB');
            $this->model = new Mapper($db, $table);
        }
```

**Comment by bcosca** [comment]
That's a really bad app design if you ask me. I'm really not convinced this PR is necessary.

**Comment by JeromeGill** [comment]
Its just an example where you might need to compare tables of mappers. A Mapper object of one table has potentially an entirely separate set of properties from a Mapper object of a seperate table. Right now all you can do is compare schemas.

**Comment by bcosca** [comment]
This is what bothers me:

```
    public function __construct(Mapper $table = null){
        if('users' === $table->table()){
            parent::__construct($table);
        } else {
            parent::__construct('users');
        }
    }
```

Don't you think it's a waste of CPU cycles?

**Comment by JeromeGill** [comment]
Yeah, I do, however it isn't as heavy as having to call;
'''
          $f3 = \Base::instance();
            $db = $f3->get('DB');
            $this->model = new Mapper($db, $table);
'''
Followed by hitting the DB to load a particular entry when a suitable mapper object may be available already by some other means. F3 provides an awful lot of methods that provide Mapper objects and plenty of scope for passing them around.

**Comment by bcosca** [comment]
First, I don't understand why a method/constructor can have an argument that can change context: string or Mapper, because this is where the confusion begins. Is it a table? Is it a mapper? Is it a service? Is it null? The complexity at the argument level is also going to be a documentation nightmare for your app. Will this survive the next version?

**Comment by JeromeGill** [comment]
It's ok. It's a bad example.

Say I wanted to write a generator to create migrations from fatfree. (I'd like to do this :D)

I could do something like 

static function createMigrations(Mapper $mapper){
    return function(){
       $model = $mapper
        $schema = $this->model->schema();
      $table = $this->model->table();
       for each($schema as foo => bar) {
           generate some migrations for $table.".".$foo 
      }
   }
}

Alternatively you need to pass in or even instantiate a DB connection within the method and create the mapper there but then the method either no longer becomes DB agnostic, gets much longer, or your still passing two arguments anyway. 

Basically the mapper has a lot of great reusable stuff you can levy for this kind of thing but having no access to the table name just forces arbitrary work a rounds when there is really no need in my opinion.

**Comment by ikkez** [comment]
why create migrations from mappers? This sample looks like an extra round trip. On instantiation of the mappers, you already have the table name. Go there and use the DB\SQL class directly, which also has the schema function (mapper->schema just forwards it to db->schema), see http://fatfreeframework.com/sql#schema
So actually it should be
```php
createMigrations(string $tableName){
$db = \Base::instance()->get('DB'); // your DB object
$schema = $db->schema($tableName);
foreach($schema as $field => $config)
...
}
```

**Comment by JeromeGill** [comment]
Migrations aren't the only thing you would generate. What about fixtures? Sure you could populate random data but real data, or modified real data is always better.

**Comment by JeromeGill** [comment]
At the end of the day, its the table name. It's just a convenience method. There is always going to be another way of dealing with it being unavailable (with varying degrees of convenience) when it matters and most of the time it doesn't.

It's basically needed only for defensive programming where you want to make sure the mapper is the correct one, perhaps in a unit test for a repository that manages multiple tables, for example. If you really don't want it, it's honestly ok! I just would have found it handy a few times and thought others would too :smile:

**Comment by bcosca** [comment]
Make sure you make an identical PR for fatfree-core. I should have rejected this in the first place for being the wrong repo to post PRs to.

**Comment by JeromeGill** [comment]
Whey! Thank you!

Already done.
https://github.com/bcosca/fatfree-core/pull/19
